File: python/01570_dot_product_of _two_sparse_vectors.py
# Time complexity - O (n)
# Space complexity - O (n)
class SparseVector:
    def __init__(self, nums):
        self.non_zero_elements = {i: num for i, num in enumerate(nums) if num != 0}

    def dotProduct(self, vec):
        result = 0
        for index, value in self.non_zero_elements.items():
            if index in vec.non_zero_elements:
                result += value * vec.non_zero_elements[index]
        return result

File: java/01570_dot_product_of _two_sparse_vectors.java
import java.util.HashMap;
import java.util.Map;

// Time complexity - O (n)
// Space complexity - O (n)
class SparseVector {
    private Map<Integer, Integer> nonZeroElements = new HashMap<>();

    public SparseVector(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                nonZeroElements.put(i, nums[i]);
            }
        }
    }

    public int dotProduct(SparseVector vec) {
        int result = 0;
        for (Map.Entry<Integer, Integer> entry : nonZeroElements.entrySet()) {
            int index = entry.getKey();
            int value = entry.getValue();
            if (vec.nonZeroElements.containsKey(index)) {
                result += value * vec.nonZeroElements.get(index);
            }
        }
        return result;
    }
}
